randovania/resolver/state.py

```diff
@@ -108,6 +114,25 @@ def game_state_debug_string(self) -> str:
         """A string that represents the game state for purpose of resolver and generator logs."""
         return self.damage_state.debug_string(self.resources)
 
+    def _advance_to(
+        self,
+        new_resources: ResourceCollection,
+        new_collected_resource_nodes: tuple[ResourceNode, ...],
+        damage_state: DamageState,
+        patches: GamePatches,
+    ) -> Self:
+        return State(
```

Comment:
Annotate the class with `@final` such that mypy knows that `Self` is really always `State`.